controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,287 @@
+	if err != nil {
+		r.Log.Error(err, "failed to get storage cluster")
+		return reconcile.Result{}, err
+	}
+
+	// list the storagecluster deviceset storageclass
+	err = r.detectLsoStorageclass(ctx, storageCluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// check if the ceph cluster is healthy
+	notHealthy, err := r.checkIfCephClusterIsNotHealthy(ctx, storageCluster)
+	if err != nil {
+		r.Log.Error(err, "failed to check if ceph cluster is healthy")
+		return reconcile.Result{}, err
+	}
+	if notHealthy {
+		err = fmt.Errorf("ceph cluster is not healthy")
+		r.Log.Error(err, "scaling cannot be performed")
+		// return nil error as scraper will retry if needed
+		return reconcile.Result{}, nil
+	}
+
+	// get the osd usage from the sync map
+	usage, ok := r.SyncMap.Load(storageAutoScaler.Spec.DeviceClass)
+	if !ok {
+		r.Log.Info("osd usage not found for device class in sync map", "device class", storageAutoScaler.Spec.DeviceClass)
+		// return nil error as scraper will retry if needed
+		return reconcile.Result{}, nil
+	}
+
+	if usage == nil {
+		r.Log.Info("osd usage is nil for device class in sync map", "device class", storageAutoScaler.Spec.DeviceClass)
+		// return nil error as scraper will retry if needed
+		return reconcile.Result{}, nil
+	}
+
+	if !checkIfScalingRequired(usage, storageAutoScaler.Spec.StorageScalingThresholdPercent) {
+		r.Log.Info("osd usage is less than the threshold", "device class", storageAutoScaler.Spec.DeviceClass, "usage", usage.(float64)*100)
+		return reconcile.Result{}, nil
+	}
+	// scaling is required
+	r.Log.Info("scaling is required", "device class", storageAutoScaler.Spec.DeviceClass)
+
+	// calculate the expectedStorageCapacity
+	startOsdSize, expectedOsdSize, startOsdCount, expectedOsdCount, startStorageCapacity, expectedStorageCapacity := calculateExpectedOsdSizeAndCount(storageCluster, storageAutoScaler)
+
+	if expectedStorageCapacity.Cmp(storageAutoScaler.Spec.StorageCapacityLimit) > 0 {
+		r.Log.Info("storage capacity limit reached")
+		originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+		// update the status
+		bool := true
+		storageAutoScaler.Status.StorageCapacityLimitReached = &bool
+		err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// update the status
+	originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+	storageAutoScaler.Status.Phase = "InProgress"
+	storageAutoScaler.Status.LastExpansion = &ocsv1.LastExpansionStatus{
+		StartOsdCount:           uint16(startOsdCount),
+		ExpectedOsdCount:        uint16(expectedOsdCount),
+		StartOsdSize:            startOsdSize,
+		ExpectedOsdSize:         expectedOsdSize,
+		StartStorageCapacity:    startStorageCapacity,
+		ExpectedStorageCapacity: expectedStorageCapacity,
+		StartTime:               metav1.Now(),
+		CompletionTime:          metav1.Time{},
```

Comment:
How does this show up in the CR status when not completed? Seems like the CompletionTime should instead be nullable, and remain nil until completion.

Reply:
Yes completion time shows null that correct.

Reply:
an empty `Time` struct show up as nil (or empty) in the status? if so, sounds good

Reply:
```
    lastExpansion:
      expectedOsdCount: 3
      expectedOsdSize: 200Gi
      expectedStorageCapacity: 600Gi
      startOsdCount: 3
      startOsdSize: 100Gi
      startStorageCapacity: 300Gi
      startTime: "2025-03-21T18:13:54Z"
    phase: InProgress

```